### PR TITLE
Various charm fixes.

### PR DESCRIPTION
**Pull request by alesstimec**
## Description

The what and why - include a summary of the change, describe what it does, and include relevant motivation and context.

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests
- [ ] Independent change*

### CODE REVIEW

File: scripts/lxd-snap-build.sh [hunk 1]
```diff
@@ -40,7 +40,7 @@ if [ -n "${http_proxy:-}" ]; then
 	lxd_exec_ubuntu git config --global http.proxy ${http_proxy:-}
 fi
 lxd_exec snap install snapcraft --classic
-lxc file push --uid 1000 --gid 1000 --mode 600 ${NETRC:-$HOME/.netrc} $container/home/ubuntu/.netrc
+lxc file push --uid 1000 --gid 1000 --mode 600 ${NETRC:-/home/alesstimec/.netrc} $container/home/ubuntu/.netrc
```
Comment (andrei-polukhin) [inline scripts/lxd-snap-build.sh]: Should we revert this change?

File: charms/candid/src/charm.py [hunk 2]
```diff
@@ -113,28 +126,43 @@ def _start(self, event: StartEvent):
...
-        if self._stored.db_uri:
+        if self._state.db_uri:
             config_values["storage"] = {
                 "type": "postgres",
-                "connection-string": self._stored.db_uri,
+                "connection-string": self._state.db_uri,
             }
         else:
-            logging.info("db_uri not stored")
+            logging.info("missing postgres relation")
```
Comment (kian99) [inline charms/candid/src/charm.py]: We should enter a blocked state here and return

File: charms/candid/src/charm.py [hunk 3]
```diff
@@ -113,28 +126,43 @@ def _start(self, event: StartEvent):
...
 
-        logging.info("setting config values {}".format(config_values))
+        relation = self.model.get_relation("candid")
+        if relation:
+            peers = []
```
Comment (kian99) [inline charms/candid/src/charm.py:141]: The scoping of this variable is weird. It's only declared inside this `if` but we don't assign it to something outside this scope.

Comment (kian99) [reply charms/candid/src/charm.py:141]: I.e. it doesn't look like it's actually being used.

**Comment by mina1460** [comment]
the linter configuration is not set to check new changes only, it will fail because it is trying to lint the entire repo. I did the change required to lint new changes only in livepatch-client. 

I can do it here if you would like

File: charms/candid/src/charm.py (hunk 2 above)
Comment (kian99) [reply charms/candid/src/charm.py]: Though granted lower down if the DB connection string is removed we call this function and if we were to return early here we wouldn't stop the Candid snap so it's a bit tricky.

File: charms/candid/src/charm.py [hunk 4]
```diff
@@ -234,47 +258,57 @@ def _on_master_changed(self, event: pgsql.MasterChangedEvent) -> None:
             "Updating database configuration...", WaitingStatus
         )
 
-        uri = event.master.uri
-        if not uri:
-            uri = ""
-        logging.info("database uri {}".format(uri))
+        if not event.master or not event.master.uri:
+            logging.debug("removing database connection string")
+            del self._state.db_uri
```
Comment (kian99) [inline charms/candid/src/charm.py:243]: Maybe in this case calling the stop hook and returning might be good?

**Review by kian99 — APPROVED** [review]

**Comment by mina1460** [comment]
rebased to run the actions again

File: charms/candid/src/state.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling charm state."""
+
+import functools
+import json
+
+
+def requires_state_setter(func):
```
Comment (babakks) [inline charms/candid/src/state.py:10]: Nice replacement for repeated `state.ready` checks.

**Review by babakks — APPROVED** [review]

File: charms/candid/src/charm.py (hunk 3 above)
Comment (alesstimec) [reply charms/candid/src/charm.py:141]: i append to it on L161 and use it on L163?

File: charms/candid/src/charm.py (hunk 2 above)
Comment (alesstimec) [reply charms/candid/src/charm.py]: removed this check as it is not needed.. code calls self._check_config(event) which will return False if postgres connection string is missing and go into Waiting status

File: charms/candid/src/charm.py (hunk 3 above)
Comment (kian99) [reply charms/candid/src/charm.py:141]: Ah don't know how I missed that. Looks good.

File: charms/candid/src/charm.py (hunk 4 above)
Comment (alesstimec) [reply charms/candid/src/charm.py:243]: this hook calls _update_config_and_restart which will now stop the service should any config value be missing

File: charms/candid/src/state.py (hunk 5 above)
Comment (alesstimec) [reply charms/candid/src/state.py:10]: yes. but we have to be careful using it.. it is only useful when you're setting data in application's data bucket, which requires the unit to be leader.. if you're setting data in unit's data bucked, you don't need this check
